### PR TITLE
add cleanup functionality

### PR DESCRIPTION
**Pull request by rogpeppe**

### CODE REVIEW

File: patch.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+package quicktest
```
Comment (frankban) [inline patch.go]: Add license header: `// Licensed under the MIT license, see LICENCE file for details.`

File: patch_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package quicktest_test
```
Comment (frankban) [inline patch_test.go]: License header.

File: patch_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"testing"
+
+	qt "github.com/frankban/quicktest"
+)
+
+func TestSetInt(t *testing.T) {
```
Comment (frankban) [inline patch_test.go]: s/TestSetInt/TestPatchInt/

File: patch_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	c := qt.New(t)
+	i := 99
+	c.Patch(&i, 88)
+	c.Assert(i, qt.Equals, 88)
+	c.Cleanup()
+	c.Assert(i, qt.Equals, 99)
+}
+
+func TestSetError(t *testing.T) {
```
Comment (frankban) [inline patch_test.go]: TestPatchError

File: patch_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	oldErr := errors.New("foo")
+	newErr := errors.New("bar")
+	err := oldErr
+	c.Patch(&err, newErr)
+	c.Assert(err, qt.Equals, newErr)
+	c.Cleanup()
+	c.Assert(err, qt.Equals, oldErr)
+}
+
+func TestSetErrorToNil(t *testing.T) {
```
Comment (frankban) [inline patch_test.go]: TestPatchErrorToNil

File: patch_test.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	c := qt.New(t)
+	oldErr := errors.New("foo")
+	err := oldErr
+	c.Patch(&err, nil)
+	c.Assert(err, qt.Equals, nil)
+	c.Cleanup()
+	c.Assert(err, qt.Equals, oldErr)
+}
+
+func TestSetMapToNil(t *testing.T) {
```
Comment (frankban) [inline patch_test.go]: TestPatchMapToNil

File: patch_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	c := qt.New(t)
+	oldMap := map[string]int{"foo": 1234}
+	m := oldMap
+	c.Patch(&m, nil)
+	c.Assert(m, qt.IsNil)
+	c.Cleanup()
+	c.Assert(m, qt.DeepEquals, oldMap)
+}
+
+func TestSetPanicsWhenNotAssignable(t *testing.T) {
```
Comment (frankban) [inline patch_test.go]: TestPatchPanicsWhenNotAssignable

File: patch_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+	c := qt.New(t)
+	i := 99
+	type otherInt int
+	c.Assert(func() { c.Patch(&i, otherInt(88)) }, qt.PanicMatches, `reflect\.Set: value of type quicktest_test\.otherInt is not assignable to type int`)
+}
+
+func TestSetenv(t *testing.T) {
+	c := qt.New(t)
+	const envName = "TESTING_PATCH_ENVIRONMENT"
+	// remember the old value, and set it to something we can check
```
Comment (frankban) [inline patch_test.go]: s/remember/Remember/ amd final dot.

File: patch_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	c.Assert(c, qt.Not(qt.Equals), "")
+	info, err := os.Stat(dir)
+	c.Assert(err, qt.Equals, nil)
+	c.Assert(info.IsDir(), qt.Equals, true)
+	f, err := os.Create(filepath.Join(dir, "hello"))
+	c.Assert(err, qt.Equals, nil)
+	f.Close()
+	c.Cleanup()
+	_, err = os.Stat(dir)
+	c.Assert(err, qt.Not(qt.IsNil))
```
Comment (frankban) [inline patch_test.go]: `c.Assert(os.IsNotExist(err), qt.Equals, true)` ?

File: quicktest_test.go [hunk 10]
```diff
@@ -149,6 +149,18 @@ func TestCRunPanic(t *testing.T) {
...
+func TestCRunCleanup(t *testing.T) {
+	c := qt.New(&testingT{})
+	outerClean := 0
+	innerClean := 0
+	c.AddCleanup(func() { outerClean++ })
+	c.Run("x", func(c *qt.C) {
+		c.AddCleanup(func() { innerClean++ })
+	})
+	c.Assert(innerClean, qt.Equals, 1)
+	c.Assert(outerClean, qt.Equals, 0)
```
Comment (frankban) [inline quicktest_test.go]: Very nice.

File: quicktest_test.go [hunk 11]
```diff
@@ -149,6 +149,18 @@ func TestCRunPanic(t *testing.T) {
...
+}
+
 func checkResult(t *testing.T, ok bool, got, want string) {
 	if want != "" {
 		assertPrefix(t, got, "\n"+want)
```
Comment (frankban) [inline quicktest_test.go]: We should add tests for AddCleanup and Cleanup, as they seem to be exercised only indirectly.

File: patch_test.go (hunk 9 above)
Comment (rogpeppe) [reply patch_test.go]: Yeah, maybe. I miss Satisfies - maybe we could add that.

File: quicktest_test.go (hunk 11 above)
Comment (rogpeppe) [reply quicktest_test.go]: done

File: patch_test.go (hunk 8 above)
Comment (rogpeppe) [reply patch_test.go]: i removed the comments; they weren't adding much.

File: patch_test.go (hunk 7 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch_test.go (hunk 6 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch_test.go (hunk 5 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch_test.go (hunk 4 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch_test.go (hunk 3 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch_test.go (hunk 2 above)
Comment (rogpeppe) [reply patch_test.go]: done

File: patch.go (hunk 1 above)
Comment (rogpeppe) [reply patch.go]: done

**Pull request merged**